src/_pytest/pathlib.py

```
@@ -549,17 +549,32 @@ def resolve_package_path(path: Path) -> Optional[Path]:
 
 
 def visit(
-    path: str, recurse: Callable[["os.DirEntry[str]"], bool]
+    path: str,
+    recurse: Callable[["os.DirEntry[str]"], bool],
+    *,
+    _seen: Optional[Set[str]] = None,
 ) -> Iterator["os.DirEntry[str]"]:
     """Walk a directory recursively, in breadth-first order.
 
     Entries at each directory level are sorted.
     """
+    if _seen is None:
+        _seen = set()
+
+    real_path = os.path.realpath(path)
+    if real_path in _seen:
+        return
+    _seen.add(real_path)
+
     entries = sorted(os.scandir(path), key=lambda entry: entry.name)
     yield from entries
     for entry in entries:
-        if entry.is_dir(follow_symlinks=False) and recurse(entry):
-            yield from visit(entry.path, recurse)
+        try:
+            is_directory = entry.is_dir(follow_symlinks=True)
+        except OSError:
```

[major] Catching a bare  here will now silently swallow permission problems or other filesystem errors, which used to propagate and fail loudly. In the internal logic we only want to skip the specific cases we expect (e.g. a broken symlink or ), and let everything else surface so we notice real regressions. Please narrow the exception handling to the explicit error codes we anticipate and re-raise unexpected ones.

---

Updated the traversal guard to only skip the errno cases we expect (dangling link, loop, access denied) and let the rest raise; re-ran the symlink regression tests locally.